proto/xmtpv4/message_api/message_api.proto

```diff
@@ -0,0 +1,129 @@
+// Message API for XMTP V4
+syntax = "proto3";
+
+package xmtp.xmtpv4;
+
+import "google/api/annotations.proto";
+import "identity/associations/signature.proto";
+
+option go_package = "github.com/xmtp/proto/v3/go/xmtpv4/message_api";
+
+// Authenticated data within the MlsPrivateMessage
+message AuthenticatedData {
+  repeated uint64 last_originator_sids = 1;
+}
+
+// Replaces GroupMessageInput V1
+// To rename or not to rename?
+message ClientEnvelope {
+  // TLS serialized MlsMessageIn, which contains MlsPrivateMessage
+  bytes data = 1;
+  bytes sender_hmac = 2;
+}
+
+// Wraps client envelope with payer signature
+message PayerEnvelope {
+  bytes unsigned_client_envelope = 1; // Protobuf serialized
+  xmtp.identity.associations.RecoverableEcdsaSignature payer_signature = 2;
+}
+
+// For blockchain envelopes, the originator_sid is set by the smart contract,
+// but the originator_ns is set by the publishing node
+message UnsignedOriginatorEnvelope {
+  uint64 originator_sid = 1;
+  uint64 originator_ns = 2;
+  PayerEnvelope payer_envelope = 3;
+}
+
+// An alternative to a signature for blockchain payloads
+message BlockchainProof {
+  uint64 block_number = 1;
+  uint32 publisher_id = 2;
+}
+
+// Signed originator envelope
+message OriginatorEnvelope {
+  bytes unsigned_originator_envelope = 1; // Protobuf serialized
+  oneof proof {
+    xmtp.identity.associations.RecoverableEcdsaSignature originator_signature = 2;
+    BlockchainProof blockchain_proof = 3;
+  }
+}
+
+// Wraps originator envelope with a sequence ID assigned by the gateway
+message GatewayEnvelope {
```

Comment:
I wonder if "gateway" is the right term here. When I hear "gateway" my mind immediately associates it with being the first thing that touches the message. In this case it's really the last.

What about `ReplicaEnvelope`?